src/forms/Typeahead.jsx

```diff
@@ -15,39 +15,41 @@ class Typeahead extends React.PureComponent {
 	constructor(props) {
 		super(props);
 		this.onItemMouseEnter = this.onItemMouseEnter.bind(this);
-		this.onItemKeyUp = this.onItemKeyUp.bind(this);
+		this.stateReducer = this.stateReducer.bind(this);
 
 		this.state = {
-			actualHighlightedIndex: 0
+			highlightedIndex: -1
 		};
 	}
 
 	onItemMouseEnter (i) {
-		this.setState({actualHighlightedIndex: i});
+		this.setState({highlightedIndex: i});
 	}
 
-	onItemKeyUp(e) {
-		switch (e.key) {
-			case 'ArrowDown':
+	stateReducer(state, changes) {
+		switch (changes.type) {
+			case Downshift.stateChangeTypes.keyDownArrowDown:
 				this.setState({
```

Comment:
The `changes` argument here refers to the parts of state that will be updated in `Downshift` state change. For example....

Let's say this is the initial `state` of `Downshift`:
```
state = {
	selectedIndex: 0,
	highlightedIndex: 0,
	foo: 'bar',
};
```

Then let's say that `Downshfit` is changing state for `highlightedIndex` only, based on some user action. This is what `stateReducer` would be called with:

```
stateReducer = (state, changes) => {
	console.dir(state); // full state object as described above
	console.dir(changes); // only the parts of state that change, { highlightedIndex: <updated value> }

	// stateReducer returns the parts of state to update
	// to use as an argument for a setState in Downshift
	return changes;
}
```

Downshift is essentially doing this in calls to its `setState`:

`setState(this.props.stateReducer(this.state, { <state changes> }))`

----

Rather than fixing this issue only in `Typeahead` internal state, `stateReducer` provides an opportunity to intercept `Downshift` state changes to amend the values as needed:

```
return { ...changes, { highlightedIndex: <our calculatedIndex> } }
```

This will ensure that any management of state in `Typeahead` is in sync with `Downshift` state.

Reply:
This may also solve any weirdness between `highlightedIndex` and `selectedIndex` in Downshift

Reply:
I originally did it this way, but then I _always_ need to use `highlightedIndex` to check if we should be adding the className `typeahead-item--isActive`.
This is a problem because we can't rely on `highlightedIndex` to be correct when navigating with a cursor. In the venue search, Typeahead doesn't even do an internal state change when using the cursor on some of the dropdown items, so I can't even bring that interaction into `stateReducer` as well

Reply:
So `Typeahead` can keep the internal state you're using for `highlightedIndex`, but we should also try passing that state property to `Downshift` in the return value to correct its internal state of highlighted item

Reply:
Ahhh, I understand. Ok, I'll do that 👍